proposals/0055-csp.md

```diff
@@ -0,0 +1,443 @@
+### Configuration APIs
+
+New configurations will be available in case directives and/or hashes can be applied to all pages
+
+```js
+export default defineConfig({
+    security: {
+        csp: {
+            // or SHA-384, or SHA-256 (default)
+            algorithm: "SHA-512",
+            // Additional directives added to the final content
+            directives: [
+                "default-src 'self'", 
+                "img-src 'self'",
+            ],
+            // customise the `style-src` directive
+            styleDirective: {
+                // A list of resources that will override the default resources of Astro
+                resources: [
+                    "self",
+                    "https://css.cdn.example.com",
+                ],
+                // List of hashes for `style-src`. 
+                hashes: [
+                    "sha512-hash1",
+                    "sha256-hash1",
+                    "sha384-hash1",
+                ]
+            },
+            // customise the `style-src` directive
+            scriptDirective: {
+                // A list of resources that will override the default resources of Astro
+                resources: [
+                    "self",
+                    "https://scripts.cdn.example.com",
+                ],
+                // List of hashes for `script-src`. 
+                hashes: [
+                    "sha512-hash1",
+                    "sha256-hash1",
+                    "sha384-hash1",
+                    
+                ]
+            }
+        }
+    }
+})
+```
+
+#### Provide additional directives
+
+User can add additional directives to control other resources other than `script-src` and `style-src`.
+
+These directives is a list of strings. Each string has this form
+
+```ts
+type CspDirective = `${AllowedDirectives} ${string}`;
+```
+
+Where `AllowedDirectives` is a list of the following directives:
+
+```ts
+export const ALLOWED_DIRECTIVES = [
+	'base-uri',
+	'child-src',
+	'connect-src',
+	'default-src',
+	'fenced-frame-src',
+	'font-src',
+	'form-action',
+	'frame-ancestors',
+	'frame-src',
+	'img-src',
+	'manifest-src',
+	'media-src',
+	'object-src',
+	'referrer',
+	'report-to',
+	'require-trusted-types-for',
+	'sandbox',
+	'trusted-types',
+	'upgrade-insecure-requests',
+	'worker-src',
+] as const;
+```
+
+Here's an example where we add `default-src` directive amd `img-src` directive:
+
+```js
+export default defineConfig({
+    security: {
+        csp: {
+            // Additional directives added to the final content
+            directives: [
+                "default-src 'self'",
+                "img-src 'self' 'https://images.cdn.example.com'"
+            ]
+        }
+    }
+})
+```
+
+Will result in something similar to the following (the result is formatted to allow better readability):
+
+```html
+<meta 
+  http-equiv="Content-Security-Policy"
+  content="
+    default-src 'self';
+    img-src 'self' 'https://images.cdn.example.com';
```

Comment:
This would be invalid. Ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy#host-source

```suggestion
    img-src 'self' https://images.cdn.example.com;
```

<img width="786" alt="Screenshot 2025-06-06 at 00 19 22" src="https://github.com/user-attachments/assets/799bcbc4-01b0-4515-b9fe-49afb28a2562" />